### ISSUE TITLE
Task 4: Implement ensemble methods and model combination strategies

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ts-rs#50]
# Task 4: Implement Ensemble Methods and Model Combination Strategies

**Related to Epic**: #46  
**Depends on**: #47, #48, #49

## Objective
Implement sophisticated ensemble methods that combine multiple ML models (LSTM, GRU, Transformers) with traditional forecasting methods to achieve superior prediction accuracy and robustness.

## Scope
- Create `src/ml/ensemble.rs` with ensemble implementations
- Implement voting, stacking, and dynamic weighting strategies
- Add model diversity metrics and intelligent combination logic
- Integrate with existing forecasting methods (ARIMA, ETS, etc.)

### Core Functions to Implement
```rust
// Main ensemble interface
pub fn create_ensemble_forecaster(
    models: Vec<Box<dyn ForecastMethod>>,
    config: &EnsembleConfig,
) -> Result<EnsembleForecaster, MLError>;

// Training and prediction
pub fn train_ensemble(
    ensemble: &mut EnsembleForecaster,
    training_data: &TimeSeries,
    validation_data: &TimeSeries,
) -> Result<EnsembleTrainingResult, MLError>;

// Dynamic weight optimization
pub fn optimize_ensemble_weights(
    models: &[Box<dyn ForecastMethod>],
    validation_data: &TimeSeries,
    optimization_method: WeightOptimizationMethod,
) -> Result<Vec<f64>, MLError>;
```

### Data Structures
```rust
pub struct EnsembleForecaster {
    pub models: Vec<Box<dyn ForecastMethod>>,
    pub combination_strategy: CombinationStrategy,
    pub model_weights: Vec<f64>,
    pub diversity_metrics: DiversityMetrics,
    pub performance_history: PerformanceHistory,
    pub dynamic_weighting: bool,
}

pub struct EnsembleConfig {
    pub combination_strategy: CombinationStrategy,
    pub weight_optimization: WeightOptimizationMethod,
    pub diversity_threshold: f64,
    pub performance_window: usize,
    pub dynamic_adaptation: bool,
    pub outlier_detection: bool,
    pub confidence_weighting: bool,
}

pub enum CombinationStrategy {
    SimpleAverage,
    WeightedAverage { weights: Vec<f64> },
    Voting { threshold: f64 },
    Stacking { meta_model: Box<dyn ForecastMethod> },
    DynamicWeighting { adaptation_rate: f64 },
    PerformanceBased { lookback_window: usize },
    ConfidenceWeighted,
    BayesianModelAveraging,
}

pub enum WeightOptimizationMethod {
    EqualWeights,
    InverseVariance,
    ReturnsBasedOptimization,
    ConvexOptimization,
    GeneticAlgorithm,
    BayesianOptimization,
    OnlineGradientDescent,
}
```

### 1. Basic Ensemble Methods
```rust
impl EnsembleForecaster {
    // Simple averaging of predictions
    pub fn simple_average_forecast(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        let predictions: Vec<_> = self.models.iter()
            .map(|model| model.forecast(data, horizon))
            .collect::<Result<Vec<_>, _>>()?;
            
        let averaged = average_predictions(&predictions);
        Ok(averaged)
    }
    
    // Weighted combination based on historical performance
    pub fn weighted_average_forecast(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        let predictions: Vec<_> = self.models.iter()
            .map(|model| model.forecast(data, horizon))
            .collect::<Result<Vec<_>, _>>()?;
            
        let weighted = weighted_combination(&predictions, &self.model_weights);
        Ok(weighted)
    }
}
```

### 2. Advanced Stacking
```rust
pub struct StackingEnsemble {
    pub base_models: Vec<Box<dyn ForecastMethod>>,
    pub meta_model: Box<dyn ForecastMethod>,
    pub cross_validation_folds: usize,
    pub feature_engineering: MetaFeatureConfig,
}

pub struct MetaFeatureConfig {
    pub include_predictions: bool,
    pub include_confidence: bool,
    pub include_model_agreement: bool,
    pub include_prediction_variance: bool,
    pub include_temporal_features: bool,
    pub include_error_patterns: bool,
}

impl StackingEnsemble {
    pub fn train_meta_model(&mut self, data: &TimeSeries) -> Result<(), MLError> {
        // Generate out-of-fold predictions for meta-model training
        let meta_features = self.generate_meta_features(data)?;
        self.meta_model.train(&meta_features)?;
        Ok(())
    }
    
    fn generate_meta_features(&self, data: &TimeSeries) -> Result<TimeSeries, MLError> {
        // Create meta-features from base model predictions and characteristics
    }
}
```

### 3. Dynamic Weight Adaptation
```rust
pub struct DynamicWeightManager {
    pub adaptation_method: AdaptationMethod,
    pub performance_window: usize,
    pub learning_rate: f64,
    pub weight_bounds: (f64, f64),
    pub regularization: f64,
}

pub enum AdaptationMethod {
    ExponentialSmoothing { alpha: f64 },
    KalmanFilter { process_noise: f64, measurement_noise: f64 },
    OnlineGradient { learning_rate: f64 },
    ReinforcementLearning { exploration_rate: f64 },
}

impl DynamicWeightManager {
    pub fn update_weights(
        &mut self,
        current_weights: &[f64],
        model_errors: &[f64],
        confidence_scores: &[f64],
    ) -> Vec<f64> {
        match self.adaptation_method {
            AdaptationMethod::ExponentialSmoothing { alpha } => {
                self.exponential_smoothing_update(current_weights, model_errors, alpha)
            }
            AdaptationMethod::OnlineGradient { learning_rate } => {
                self.gradient_based_update(current_weights, model_errors, learning_rate)
            }
            // ... other adaptation methods
        }
    }
}
```

### 4. Model Diversity Analysis
```rust
pub struct DiversityMetrics {
    pub prediction_correlation: f64,
    pub error_correlation: f64,
    pub disagreement_measure: f64,
    pub entropy_based_diversity: f64,
    pub temporal_diversity: Vec<f64>,
}

impl DiversityMetrics {
    pub fn compute_diversity(
        predictions: &[Vec<f64>],
        ground_truth: &[f64],
    ) -> Self {
        let prediction_correlation = Self::compute_prediction_correlation(predictions);
        let error_correlation = Self::compute_error_correlation(predictions, ground_truth);
        let disagreement = Self::compute_disagreement_measure(predictions);
        let entropy_diversity = Self::compute_entropy_diversity(predictions);
        let temporal_diversity = Self::compute_temporal_diversity(predictions);
        
        Self {
            prediction_correlation,
            error_correlation,
            disagreement_measure: disagreement,
            entropy_based_diversity: entropy_diversity,
            temporal_diversity,
        }
    }
    
    pub fn is_diverse_enough(&self, threshold: f64) -> bool {
        self.disagreement_measure > threshold && 
        self.prediction_correlation < (1.0 - threshold)
    }
}
```

### 5. Confidence-Based Weighting
```rust
pub struct ConfidenceWeightedEnsemble {
    pub models: Vec<Box<dyn ForecastMethod>>,
    pub confidence_calibration: ConfidenceCalibration,
    pub min_confidence_threshold: f64,
    pub confidence_smoothing: f64,
}

impl ConfidenceWeightedEnsemble {
    pub fn forecast_with_confidence(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        let mut predictions = Vec::new();
        let mut confidences = Vec::new();
        
        for model in &self.models {
            let result = model.forecast(data, horizon)?;
            predictions.push(result.predictions);
            
            // Extract or compute confidence scores
            let confidence = result.confidence_intervals
                .map(|ci| self.confidence_from_intervals(&ci))
                .unwrap_or_else(|| self.estimate_confidence(model, data));
            confidences.push(confidence);
        }
        
        let weighted_prediction = self.combine_with_confidence(&predictions, &confidences);
        Ok(weighted_prediction)
    }
}
```

### 1. Traditional Model Integration
```rust
// Wrapper for traditional forecasting methods
pub struct TraditionalModelWrapper {
    pub method: ForecastMethod,
    pub uncertainty_estimation: UncertaintyMethod,
}

impl TraditionalModelWrapper {
    pub fn wrap_arima(config: ArimaConfig) -> Self {
        // Wrap ARIMA with uncertainty estimation
    }
    
    pub fn wrap_ets(config: ETSConfig) -> Self {
        // Wrap Exponential Smoothing with uncertainty estimation
    }
}
```

### 2. Heterogeneous Ensemble Support
```rust
pub struct HeterogeneousEnsemble {
    pub neural_models: Vec<Box<dyn NeuralForecastMethod>>,
    pub traditional_models: Vec<Box<dyn ForecastMethod>>,
    pub combination_weights: CombinationWeights,
    pub model_selection_criteria: SelectionCriteria,
}

pub struct CombinationWeights {
    pub neural_weight: f64,
    pub traditional_weight: f64,
    pub adaptive: bool,
    pub selection_method: ModelSelectionMethod,
}
```

### 1. Parallel Model Training
```rust
impl EnsembleForecaster {
    pub async fn train_models_parallel(&mut self, data: &TimeSeries) -> Result<(), MLError> {
        let training_tasks: Vec<_> = self.models.iter_mut()
            .map(|model| tokio::spawn(model.train(data.clone())))
            .collect();
            
        for task in training_tasks {
            task.await??;
        }
        
        Ok(())
    }
}
```

### 2. Prediction Caching
```rust
pub struct EnsemblePredictionCache {
    pub cache: HashMap<DataHash, Vec<ForecastResult>>,
    pub max_size: usize,
    pub ttl: Duration,
}
```

## Acceptance Criteria
- [ ] Ensemble methods consistently outperform individual models
- [ ] Dynamic weighting adapts effectively to changing conditions
- [ ] Stacking ensemble shows improved accuracy over simple averaging
- [ ] Model diversity metrics correctly identify diverse vs redundant models
- [ ] Confidence-based weighting improves prediction reliability
- [ ] Integration with all existing forecasting methods works seamlessly
- [ ] Parallel training and prediction provides performance benefits
- [ ] Comprehensive test suite validates all combination strategies

## Test Cases Required
- Simple ensemble with 2-3 models showing improved accuracy
- Dynamic weight adaptation during regime changes
- Stacking ensemble outperforming base models
- Model diversity analysis with known diverse/similar models
- Confidence calibration accuracy
- Performance comparison: ensemble vs best individual model

## Dependencies
- Task 1, 2, 3 results for neural network models
- Existing `forecasting` module for traditional methods
- `rayon` for parallel processing
- `tokio` for async training
- Optimization libraries for weight optimization

## Estimated Time
15-18 hours

## Next Steps
After completion, enables:
- Task 5: AutoML for automatic ensemble configuration
- Task 6: Advanced interpretability across ensemble models
- Task 7: Real-time ensemble adaptation and monitoring
